orc/src/main/java/org/apache/iceberg/orc/OrcMetrics.java

```diff
@@ -19,51 +19,211 @@
+                  EPOCH.plus(minStats.getTime(), ChronoUnit.MILLIS).toLocalDate())))
+          .orElse(null);
+    } else if (columnStats instanceof TimestampColumnStatistics) {
+      TimestampColumnStatistics tColStats = (TimestampColumnStatistics) columnStats;
+      Timestamp minValue = ((Types.TimestampType) column.type()).shouldAdjustToUTC() ?
+          tColStats.getMinimum() : tColStats.getMinimumUTC();
+      min = Optional.ofNullable(minValue)
+          .map(v -> Conversions.toByteBuffer(column.type(), toMicros(v)))
+          .orElse(null);
+    } else if (columnStats instanceof BooleanColumnStatistics) {
+      BooleanColumnStatistics booleanStats = (BooleanColumnStatistics) columnStats;
+      min = booleanStats.getFalseCount() > 0 ?
+          Conversions.toByteBuffer(column.type(), false) :
+          Conversions.toByteBuffer(column.type(), true);
+    }
+    return Optional.ofNullable(min);
   }
+
+  private static Optional<ByteBuffer> fromOrcMax(Types.NestedField column,
+                                                 ColumnStatistics columnStats) {
+    ByteBuffer max = null;
+    if (columnStats instanceof IntegerColumnStatistics) {
+      IntegerColumnStatistics intColStats = (IntegerColumnStatistics) columnStats;
+      max = column.type().typeId() == Type.TypeID.INTEGER ?
+          Conversions.toByteBuffer(column.type(), (int) intColStats.getMaximum()) :
+          Conversions.toByteBuffer(column.type(), intColStats.getMaximum());
+    } else if (columnStats instanceof DoubleColumnStatistics) {
+      double maxVal = ((DoubleColumnStatistics) columnStats).getMaximum();
+      max = column.type().typeId() == Type.TypeID.DOUBLE ?
+          Conversions.toByteBuffer(column.type(), maxVal) :
+          Conversions.toByteBuffer(column.type(), (float) maxVal);
+    } else if (columnStats instanceof StringColumnStatistics) {
+      max = Conversions.toByteBuffer(column.type(),
+          ((StringColumnStatistics) columnStats).getMaximum());
+    } else if (columnStats instanceof DecimalColumnStatistics) {
+      max = Optional
+          .ofNullable(((DecimalColumnStatistics) columnStats).getMaximum())
+          .map(maxStats -> {
+            BigDecimal maxValue = maxStats.bigDecimalValue()
+                .setScale(((Types.DecimalType) column.type()).scale());
+            return Conversions.toByteBuffer(column.type(), maxValue);
+          })
+          .orElse(null);
+    } else if (columnStats instanceof DateColumnStatistics) {
+      max = Optional.ofNullable(((DateColumnStatistics) columnStats).getMaximum())
+          .map(maxStats -> Conversions.toByteBuffer(column.type(),
+              (int) ChronoUnit.DAYS.between(EPOCH_DAY,
+                  EPOCH.plus(maxStats.getTime(), ChronoUnit.MILLIS).toLocalDate())))
+          .orElse(null);
+    } else if (columnStats instanceof TimestampColumnStatistics) {
+      TimestampColumnStatistics tColStats = (TimestampColumnStatistics) columnStats;
```

Comment:
Can we increment milliseconds for the upper bound? It should be fine for the lower bound.

Reply:
We will definitely need tests for these cases to make sure we don't have a correctness bug because of ORC-611.

Reply:
I guess incrementing the upper bound by 1ms should be okay. It might be good to verify what happens for pre-epoch timestamps as well. If they are rounded towards epoch, then we may have to decrement the lower bound by 1ms for pre epoch values.

Reply:
Agreed, we should have a test for pre-epoch bounds as well.

Reply:
I found that we do need to adjust for pre-epoch bounds, however there's a bug [ORC-342](https://issues.apache.org/jira/browse/ORC-342) that creates inaccurate timestamp/stats below epoch. I've added a test that hits the bug.

Reply:
> I found that we do need to adjust for pre-epoch bounds
After further testing I found that ORC does adjust on pre-epoch by 1 millisecond. However there's a range where the actual timestamp value is invalid, described in ORC-342. I've added another comment below related to it.